File: src/main/java/com/github/sgtsilvio/gradle/android/retrofix/transform/TypeMap.java
package com.github.sgtsilvio.gradle.android.retrofix.transform;

import javassist.ClassMap;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;

/**
 * @author Silvio Giebl
 */
public class TypeMap extends ClassMap {

    private final @NotNull HashMap<String, String> prefixMap = new HashMap<>();

    @Override
    public @Nullable String get(final @NotNull Object jvmClassName) {
        final String s = (String) jvmClassName;
        for (final Entry<String, String> entry : prefixMap.entrySet()) {
            if (s.startsWith(entry.getKey())) {
                return entry.getValue() + s.substring(entry.getKey().length());
            }
        }
        return super.get(jvmClassName);
    }

    public @Nullable String putPrefix(final @NotNull String oldPrefix, final @NotNull String newPrefix) {
        return prefixMap.put(oldPrefix, newPrefix);
    }
}

File: src/main/kotlin/com/github/sgtsilvio/gradle/android/retrofix/transform/TypeMap.kt
package com.github.sgtsilvio.gradle.android.retrofix.transform

/**
 * @author Silvio Giebl
 */
class TypeMap : HashMap<String, String>() {

    private val prefixMap = HashMap<String, String>()

    override fun get(key: String): String? {
        for ((oldPrefix, newPrefix) in prefixMap) {
            if (key.startsWith(oldPrefix)) {
                return newPrefix + key.substring(oldPrefix.length)
            }
        }
        return super.get(key)
    }

    fun putPrefix(oldPrefix: String, newPrefix: String) = prefixMap.put(oldPrefix, newPrefix)
}
